File: tests/lua/emit_ifs.ml
external val ignore : 'a -> () = "nil"
external val print : 'a -> () = "print"
external val bool : bool = "true"

let a && b = if a then b else false
let a || b = if a then true else b
let not a = if a then false else true

let () = ignore { ands = (&&), ors = ( || ), not }

let () = ignore @@ fun () ->
  if bool
  then print "L"
  else print "R"
       print "R"

let () = ignore @@ fun () ->
  if bool
  then print "L"
       print "L"
  else print "R"

let () = if bool then print "Hello" else ()

let () = if bool then () else print "Hello"

let () = ignore @@ function
  | 0 -> ()
  | _ -> print "Not 0"
         print "Extra text"

File: tests/lua/emit_ifs.lua
do
  local _print = print
  local function _amp_amp(a) return function(b) return a and b end end
  local function _bar_bar(a) return function(b) return a or b end end
  local function _not(a) return not a end
  (nil)({ ands = _amp_amp, ors = _bar_bar, ["not"] = _not });
  (nil)(function(tmp)
    if true then return _print("L") end
    _print("R")
    return _print("R")
  end);
  (nil)(function(tmp)
    if not true then return _print("R") end
    _print("L")
    return _print("L")
  end)
  if true then _print("Hello") end
  if not true then _print("Hello") end
  (nil)(function(x)
    if x == 0 then return nil end
    _print("Not 0")
    return _print("Extra text")
  end)
end
